### ISSUE TITLE
performance reduction for zipapps with many *.robot files

### ISSUE THREAD

**Issue by franzhaas** [robotframework/robotframework#4728]
Dear all,

TLDR.: I believe that usage of the invalidated_caches function is causing performance issues, slightly on regular robot installations (+25% for the used example) and hugely for zipapps (+5s for the used example)

I noticed when using a lot of robot files, with only a view tests in them which run very quickly, that the runtime in a zipapp is much slower than a regular installation.

I did have a look at what is going on there, and I noticed that __ import __ is used, not importlib.util (importlib.util apears to be the recomeded methond).

A quick and dirty experiment using importlib.util instead of __ import __ did work on my minimal example, and improved speed on the given setup and example about 5-7x for the zipapp, and about 2x for the regular install.

Running the acceptance tests triggers about 1200 fails on windows (mostly permission issues) and about 250 on linux, and does not show a significant speedup (but the fails may influence runtime significantly...)

Is this a topic worth looking into? Are there known reasons not to use importlib.util? Am I fooled by the effect that missing calculations explain improved speed?

## Some data.:

See the not-very-encouraging results of the CI run.:
https://github.com/franzhaas/robotframework/actions/runs/4645976291/jobs/8222120931

The example code used.:
https://github.com/MarketSquare/compact_testprogram_distribution

## micro benchmarks
A micro benchmark in a regular environment.:
![image](https://user-images.githubusercontent.com/11477153/230738889-dd1c253c-8b6c-4fa3-82b8-63be343ba32f.png)

A micro benchmark in a zipapp.:

![image](https://user-images.githubusercontent.com/11477153/230739010-79d64c04-8163-4b78-b252-ca7058eb2178.png)

The microbenchmarks show little to very little speedup.

However.:

A microbenchmark in a regular environment.:

![image](https://user-images.githubusercontent.com/11477153/230740083-95d55673-9830-4b58-982f-5be6e50e5af2.png)

And in a zipapp.:
![image](https://user-images.githubusercontent.com/11477153/230740102-e0a68d1e-405d-4e79-bcc7-a699d79b4ed7.png)

Therefore i now belive that the invalidate_caches function is causing the slowdown....

Best regards,
Franz

**Comment by franzhaas** [comment]
Dear all,

I followed the idea that reducing the calls to invalidate_caches improves performance. 5s is *very* slow, but avoiding to invalidate the caches is probably a good idea.

A CI run with modifications that minimize the usage of invalidate_caches
https://github.com/franzhaas/robotframework/actions/runs/4652244029/jobs/8232333488

there is a single unittest fail, and all other atests pass, I believe this is not too far of.

Still open is to get rid of the invalidate_caches completely by replacing the one remaining modify sys.path operations with for example importlib.utils.

The advantages I hope for are.:

1.  faster execution, invalidating the cache less will speed up things...
2. removing a potential race condition. Modifying the sys.path, can influence other tasks which use the import system. This is relevant when having for example a streamlit frontend to robotframework.

Best regards,
Franz

**Comment by franzhaas** [comment]
I dug into this a bit more...

This is a known issue and python community is actively discussing it.

https://github.com/python/cpython/issues/103200

I did not check it myself,  but if I understand the discussion correctly, python 3.9 and earlier are not affected, python 3.12 is planned not to be affected and whether or not fixes will be implemented for 3.10 and 3.11 is open to debate. 

To my understanding invalidated_caches  is necesseray and needs to be doing something in order for robotframework to work correctly under some conditions.

So we have the option to just sit this one out and wait for 3.12...

Best regards,
Franz

**Comment by franzhaas** [comment]
The problem at hand has been solved....

**Comment by pekkaklarck** [comment]
`importlib.invalidate_cache()` is used for a reason. Without it we couldn't import modules that are created during execution and I think it also helps if a module with same name is imported from different places. As you noticed, there are tests that break if `importlib.invalidate_cache()` is removed. I'm not sure is there a need to do it always before an import like nowadays, though. We possibly could do that only if an import fails and then retry.

The reason we use `__import__` and not `importlib` for importing is that our importing code predates `importlib`. We even had our own cache invalidation logic that was made redundant by `importlib.invalidate_cache()`. There haven't been needs to switch to `importlib` otherwise, but if there ever are bigger changes to this code it certainly makes sense.
